### PR TITLE
Optimise AssertCount for read-only collections

### PR DESCRIPTION
**Pull request by Arithmomaniac**
Here's a simple first fix for #128. It also contains all the csproj changes I would expect would need changing.

Am I the only one running into issues running the tests from within Visual Studio 2017? (They checked out from the command line, though.)

### CODE REVIEW

**Comment by Arithmomaniac** [comment]
Done.

File: MoreLinq.Test/UnenumerableList.cs [hunk 1]
```diff
@@ -10,12 +10,31 @@ namespace MoreLinq.Test
     /// method of <see cref="IEnumerable"/> - either because they should be using the indexer or because they are
     /// expected to be lazily evaluated.
     /// </summary>
+    public sealed class UnenumerableList<T> : UnenumerableListBase<T>, IList<T>
```
Comment (atifaziz) [inline MoreLinq.Test/UnenumerableList.cs]: Given the revert in c8b4de7113bd4162e1a1e0e862c0445c36a3c4af, this isn't used anymore and so should be reverted to its previous state/implementation too.

File: MoreLinq/AssertCount.cs [hunk 2]
```diff
@@ -83,12 +83,21 @@ static Exception OnAssertCountFailure(int cmp, int count)
...
             {
                 return collection.Count == count
                      ? collection
                      : From<TSource>(() => throw errorSelector(collection.Count.CompareTo(count), count));
-            }
+            };
+#if IREADONLY
+            if (source is IReadOnlyCollection<TSource> readOnlyCollection)
+            {
+                return readOnlyCollection.Count == count
```
Comment (atifaziz) [inline MoreLinq/AssertCount.cs]: This code has non-trivial duplication with the optimisation for `ICollection<>`. The two blocks should be consolidated via a local function, as in:

```c#
        static IEnumerable<TSource> AssertCountImpl<TSource>(IEnumerable<TSource> source, 
            int count, Func<int, int, Exception> errorSelector)
        {
            if (source == null) throw new ArgumentNullException(nameof(source));
            if (count < 0) throw new ArgumentOutOfRangeException(nameof(count));
            if (errorSelector == null) throw new ArgumentNullException(nameof(errorSelector));

            switch (source)
            {
                case ICollection<TSource> collection:
                    return AssertKnownCount(collection.Count);
#if IREADONLY
                case IReadOnlyCollection<TSource> readOnlyCollection:
                    return AssertKnownCount(readOnlyCollection.Count);
#endif
                default:
                    return _(); IEnumerable<TSource> _()
                    {
                        var iterations = 0;
                        foreach (var element in source)
                        {
                            iterations++;
                            if (iterations > count)
                                throw errorSelector(1, count);
                            yield return element;
                        }
                        if (iterations != count)
                            throw errorSelector(-1, count);
                    }
            }

            IEnumerable<TSource> AssertKnownCount(int actualCount) =>
                actualCount == count
                ? source
                : From<TSource>(() => throw errorSelector(actualCount.CompareTo(count), count));
        }
```

File: MoreLinq.Test/AssertCountTest.cs [hunk 3]
```diff
@@ -126,9 +126,16 @@ public void AssertCountWithMismatchingCollectionCount(int sourceCount, int count
...
-        sealed class BreakingCollection<T> : BreakingSequence<T>, ICollection<T>
+        sealed class BreakingCollection<T> : BreakingCollectionBase<T>, ICollection<T>
         {
-            public BreakingCollection(int count) => Count = count;
+            public BreakingCollection(int count) : base(count)
+            {
+            }
+        }
+
+        abstract class BreakingCollectionBase<T> : BreakingSequence<T>
```
Comment (atifaziz) [inline MoreLinq.Test/AssertCountTest.cs]: The implementation hierarchy feels a little excessive here. Why not keep things separate and simplify to the following?

```diff
diff --git a/MoreLinq.Test/AssertCountTest.cs b/MoreLinq.Test/AssertCountTest.cs
index 02ad8e1..4dc1332 100644
--- a/MoreLinq.Test/AssertCountTest.cs
+++ b/MoreLinq.Test/AssertCountTest.cs
@@ -126,16 +126,9 @@ public void AssertCountWithMismatchingCollectionCount(int sourceCount, int count
             Assert.AreEqual(e.Message, message);
         }
 
-        sealed class BreakingCollection<T> : BreakingCollectionBase<T>, ICollection<T>
+        sealed class BreakingCollection<T> : BreakingSequence<T>, ICollection<T>
         {
-            public BreakingCollection(int count) : base(count)
-            {
-            }
-        }
-
-        abstract class BreakingCollectionBase<T> : BreakingSequence<T>
-        {
-            public BreakingCollectionBase(int count) => Count = count;
+            public BreakingCollection(int count) => Count = count;
 
             public int Count { get; }
 
@@ -155,11 +148,10 @@ public void AssertCountWithReadOnlyCollectionIsLazy()
             new BreakingReadOnlyCollection<object>(5).AssertCount(0);
         }
 
-        sealed class BreakingReadOnlyCollection<T> : BreakingCollectionBase<T>, IReadOnlyCollection<T>
+        sealed class BreakingReadOnlyCollection<T> : BreakingSequence<T>, IReadOnlyCollection<T>
         {
-            public BreakingReadOnlyCollection(int count) : base(count)
-            {
-            }
+            public BreakingReadOnlyCollection(int count) => Count = count;
+            public int Count { get; }
         }
 
 #endif
```

**Comment by atifaziz** [comment]
@Arithmomaniac All looks good to me and thanks for working through the review. Do you have anything to add? Otherwise I'm happy to merge this.

**Comment by Arithmomaniac** [comment]
Just a shoutout to @Milyli for giving me the time to work on this 🙂

**Comment by atifaziz** [comment]
@Arithmomaniac Merged. Thanks to you & @Milyli indeed!

**Pull request merged**